### PR TITLE
Introducing the IConnection interface

### PR DESCRIPTION
**Pull request by moozzyk**
Unifying events

### CODE REVIEW

File: client-ts/Microsoft.AspNetCore.SignalR.Client.TS/IConnection.ts [hunk 1]
```diff
@@ -0,0 +1,10 @@
+import { DataReceived, ConnectionClosed } from "./Common"
+
+export interface IConnection {
+    start(transportName: string): Promise<void>
+    send(data: any): Promise<void>
+    stop(): void
+
+    onDataReceived: DataReceived
```
Comment (moozzyk) [inline client-ts/Microsoft.AspNetCore.SignalR.Client.TS/IConnection.ts]: missing `;`

File: client-ts/Microsoft.AspNetCore.SignalR.Client.TS/Connection.ts [hunk 2]
```diff
@@ -82,14 +81,12 @@ export class Connection {
...
-        this.dataReceivedCallback = callback;
+        if (this.onClosed) {
+            this.onClosed(error);
+        }
     }
 
-    set connectionClosed(callback: ConnectionClosed) {
-        this.connectionClosedCallback = callback;
-    }
+    onDataReceived: DataReceived
```
Comment (moozzyk) [inline client-ts/Microsoft.AspNetCore.SignalR.Client.TS/Connection.ts]: missing `;`

File: client-ts/Microsoft.AspNetCore.SignalR.Client.TS/IConnection.ts [hunk 3]
```diff
@@ -0,0 +1,10 @@
+import { DataReceived, ConnectionClosed } from "./Common"
+
+export interface IConnection {
+    start(transportName: string): Promise<void>;
```
Comment (davidfowl) [inline client-ts/Microsoft.AspNetCore.SignalR.Client.TS/IConnection.ts]: We should talk about this being a string vs enum

File: client-ts/Microsoft.AspNetCore.SignalR.Client.TS/HubConnection.ts [hunk 4]
```diff
@@ -16,24 +17,29 @@ interface InvocationResultDescriptor {
...
+    static create(url: string, queryString?: string): HubConnection {
+        return new this(new Connection(url, queryString))
+    }
 
-        this.connection.dataReceived = data => {
-            this.dataReceived(data);
+    constructor(connection: IConnection);
+    constructor(url: string, queryString?: string);
+    constructor(connectionOrUrl: IConnection | string, queryString?: string) {
+        this.connection = typeof connectionOrUrl === "string" ? new Connection(connectionOrUrl, queryString) : connectionOrUrl;
```
Comment (davidfowl) [inline client-ts/Microsoft.AspNetCore.SignalR.Client.TS/HubConnection.ts]: Union types are weird...

File: client-ts/Microsoft.AspNetCore.SignalR.Client.TS/IConnection.ts (hunk 3 above)
Comment (moozzyk) [inline client-ts/Microsoft.AspNetCore.SignalR.Client.TS/IConnection.ts]: I am planning to resurrect a community PR that replaced string with enum and possibly introduce a factory if this helps with testing.

**Pull request merged**